src/routes/notifications/v1/entities/__tests__/create-registration-v2.dto.builder.ts

```diff
@@ -0,0 +1,73 @@
+import type { UpsertSubscriptionsDto } from '@/datasources/notifications/entities/upsert-subscriptions.dto.entity';
+import { AuthPayload } from '@/domain/auth/entities/auth-payload.entity';
+import { NotificationType } from '@/domain/notifications/v2/entities/notification.entity';
+import type { RegisterDeviceDto } from '@/routes/notifications/v1/entities/register-device.dto.entity';
+import type { UpsertSubscriptionsSafesDto } from '@/routes/notifications/v2/entities/upsert-subscriptions.dto.entity';
+import type { UUID } from 'crypto';
+import { recoverMessageAddress } from 'viem';
+
+export const createV2RegisterDtoBuilder = async (
+  args: RegisterDeviceDto,
+): Promise<
+  Array<{
+    upsertSubscriptionsDto: UpsertSubscriptionsDto;
+    authPayload: AuthPayload;
+  }>
+> => {
+  const safeV2Array: Array<{
+    authPayload: AuthPayload;
+    upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+      signature: `0x${string}`;
+    };
+  }> = [];
+
+  const safesV1Registrations = args.safeRegistrations;
+
+  for (const safeV1Registration of safesV1Registrations) {
+    if (safeV1Registration.safes.length) {
+      const safeV2: {
+        authPayload: AuthPayload;
+        upsertSubscriptionsDto: UpsertSubscriptionsDto & {
+          safes: Array<UpsertSubscriptionsSafesDto>;
+          signature: `0x${string}`;
+        };
+      } = {
+        upsertSubscriptionsDto: {
+          cloudMessagingToken: args.cloudMessagingToken,
+          deviceType: args.deviceType,
+          deviceUuid: (args.uuid as UUID) || undefined,
+          safes: [],
+          signature: safeV1Registration.signatures[0] as `0x${string}`,
+        },
+        authPayload: new AuthPayload(),
+      };
+      for (const safeAddresses of safeV1Registration.safes) {
+        safeV2.upsertSubscriptionsDto.safes.push({
+          address: safeAddresses as `0x${string}`,
+          chainId: safeV1Registration.chainId,
+          notificationTypes: Object.values(NotificationType),
+        });
+      }
+      safeV2Array.push(safeV2);
+    }
+  }
+
+  for (const [index, safeV2] of safeV2Array.entries()) {
+    const safeAddresses = safeV2.upsertSubscriptionsDto.safes.map(
+      (safeV2Safes) => safeV2Safes.address,
+    );
+
+    const recoveredAddress = await recoverMessageAddress({
+      message: `gnosis-safe${args.timestamp}${args.uuid}${args.cloudMessagingToken}${safeAddresses.sort().join('')}`,
+      signature: safeV2.upsertSubscriptionsDto.signature,
+    });
+
+    safeV2.authPayload.chain_id =
+      safeV2.upsertSubscriptionsDto.safes[0].chainId;
+    safeV2.authPayload.signer_address = recoveredAddress;
+
+    safeV2Array[index].authPayload = safeV2.authPayload;
```

Comment:
It’s located within tests, so I suggest we leave it as is since its purpose is to test the creation of `v2Request`.

Reply:
I mean instead of assigning an ["empty" `AuthPayload`](https://github.com/safe-global/safe-client-gateway/blob/a28bfabbec8319b10789908bc47c2dbbb689cb42/src/routes/notifications/v1/entities/__tests__/create-registration-v2.dto.builder.ts) and updating it retrospectively. We [could achieve this in the above loop](https://github.com/safe-global/safe-client-gateway/pull/2063#discussion_r1827771631).

Reply:
Well, we need the recoveredAddress and also it's preferred to have two different separate loops than nested loops.